test/extensions/transport_sockets/tls/rsa_private_key_method_provider.h

```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include "envoy/event/dispatcher.h"
+#include "envoy/server/transport_socket_config.h"
+#include "envoy/ssl/private_key/private_key.h"
+#include "envoy/ssl/private_key/private_key_config.h"
+
+#include "common/config/utility.h"
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace PrivateKeyMethodProvider {
+
+struct RsaPrivateKeyConnectionTestOptions {
+  // Return private key method value directly without asynchronous operation.
+  bool sync_mode_{};
+
+  // The "decrypt" private key method is expected to he called.
+  bool decrypt_expected_{};
+
+  // The "sign" private key method is expected to he called.
+  bool sign_expected_{};
+
+  // Add a cryptographic error (invalid signature, incorrect decryption).
+  bool crypto_error_{};
+
+  // Return an error from the private key method.
+  bool method_error_{};
+
+  // Return an error from the private key method completion function.
+  bool async_method_error_{};
+};
+
+// An example RSA private key method provider here for testing the decrypt() and sign()
+// functionality.
+class RsaPrivateKeyConnection : public virtual Ssl::PrivateKeyConnection {
+public:
+  RsaPrivateKeyConnection(SSL* ssl, Ssl::PrivateKeyConnectionCallbacks& cb,
+                          Event::Dispatcher& dispatcher, bssl::UniquePtr<EVP_PKEY> pkey,
+                          RsaPrivateKeyConnectionTestOptions& test_options);
+  RSA* getPrivateKey() { return EVP_PKEY_get0_RSA(pkey_.get()); }
```

Comment:
I guess this is an example provider, so maybe it's intentional that you're breaking it down this far, but if you just use the `EVP_PKEY`, the API at that layer is a lot less work. You don't need to call `RSA_add_pkcs1_prefix` or assemble the PSS padding payload yourself or anything. (We usually do not recommend people assemble crypto operations from raw materials like that. It's more work and you sometimes miss out on hardening work that we can only do when we handle the whole thing.)

Reply:
Example:
https://cs.chromium.org/chromium/src/net/ssl/test_ssl_private_key.cc?rcl=a1c9e2c4a01f41cbb9af5b78feda358ca154bb1b&l=41

Reply:
Actually the first version I wrote copied the private key method sign code from the BoringSSL tests. However, it meant that the file was licensed using the BoringSSL BSD license, which would have made Envoy licensing more complex. If you scroll up this PR you can see discussion about the subject. I eventually decided to rewrite the sign function using lower-level operations to avoid the code copying.

BTW, have you considered adding external symbols for basic private key methods (RSA, ECDSA) to BoringSSL? It would seem that everyone writing their apps with private key method support needs to implement the basic SW-only private key methods for testing, before going on to write their own methods (TPM etc). Ideally these built-in BoringSSL private key methods would have some way to inject various errors for testing the application logic. The symbols could be behind some BoringSSL build flag (`--enable-private-key-method-testing`) or maybe in a separate test library.

Reply:
It seems the root cause here is not that BoringSSL isn't providing APIs, but that the PR has gone out of its way to not use the APIs we provide, at several layers. Even with the low-level ones, this should be using `RSA_sign` and `RSA_sign_pss_mgf1` so we'd take care of the padding internally.  But ultimately it really should be using EVP. Per that Chromium link, signing is just a few function calls. It's documented here, and is more-or-less the same as OpenSSL, except we don't have the pesky -1 and -2 returns to worry about.
https://commondatastorage.googleapis.com/chromium-boringssl-docs/evp.h.html

I'm not sure what adding basic private key methods would accomplish. Installing them on the `SSL_CTX` would be exactly the same as not configuring an `SSL_PRIVATE_KEY_METHOD` at all, so you're not actually testing your project's `SSL_PRIVATE_KEY_METHOD` code. Rather, what you want is to test a custom implementation of your own project's offloading interface with a software-backed implementation. For that, you do not want an `SSL_PRIVATE_KEY_METHOD` but just convenience functions to do the operations on an `EVP_PKEY`. I'm not opposed to that, but, as noted, the EVP API is already pretty close and the complexity in this PR is a bit self-inflicted.

Re bug injection, that should be done in test code, not scattered throughout production code where it risks real bugs. If you want an operation to fail, just install something that fails. If you want it to give the wrong answer, give the wrong answer.